File: backend/src/main/java/com/stockanalysis/backtest/BacktestEngine.java
package com.stockanalysis.backtest;

import com.stockanalysis.backtest.spec.CapitalMode;
import com.stockanalysis.backtest.spec.CapitalSpec;
import com.stockanalysis.backtest.spec.ConditionGroup;
import com.stockanalysis.backtest.spec.ExitSpec;
import com.stockanalysis.backtest.spec.PremarketSpec;
import com.stockanalysis.backtest.spec.StrategySpec;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * Framework-independent intraday (scalping) backtest engine.
 *
 * <p>Single mode: one long position, 1 unit, entry at the close of the signal bar
 * (no look-ahead). Exit priority per held bar: stop-loss → take-profit → signal → time →
 * day-end → end-of-data; take-profit/stop-loss are intrabar via high/low, stop-loss winning
 * ties (conservative). Their percentages are re-resolved on every held bar against the bar's own
 * wall-clock time, so an {@link ExitSpec} time band can move the levels mid-position.
 *
 * <p>ETF pre-market mode ({@link #runEtfPremarket}): each day the futures close change over the
 * pre-market window picks leverage (up) or inverse (down); the chosen ETF is bought at the open
 * and closed by the same exit rules.
 *
 * <p>Entries and exits are decided purely on price; money is layered on afterwards by
 * {@link #applyCapital} using the spec's {@code CapitalSpec} and a {@link FeeSchedule} of per-symbol
 * commission rates, so position sizing and fees never change which trades happen.
 */
public class BacktestEngine {

    private record ExitDecision(double price, ExitReason reason) {
    }

    // ------------------------------------------------------------------ single mode

    /** Commission-free convenience overload: the price-only baseline used by tools and tests. */
    public BacktestResult run(StrategySpec spec, BarSeries series) {
        return run(spec, series, FeeSchedule.free());
    }

    public BacktestResult run(StrategySpec spec, BarSeries series, FeeSchedule fees) {
        List<Bar> bars = series.bars();
        int n = bars.size();
        List<TradeRecord> trades = new ArrayList<>();

        ConditionGroup entry = spec.getEntry();
        ExitSpec exitSpec = spec.getExit();
        ConditionGroup exitGroup = exitSpec.asGroup();

        boolean holding = false;
        int entryIndex = -1;
        double entryPrice = 0.0;
        LocalDateTime entryTs = null;

        for (int i = 0; i < n; i++) {
            Bar cur = bars.get(i);
            Bar prev = i > 0 ? bars.get(i - 1) : null;
            Bar curFut = series.hasFutures() ? series.futuresAt(cur.ts()) : null;
            Bar prevFut = (prev != null && series.hasFutures()) ? series.futuresAt(prev.ts()) : null;

            if (!holding) {
                if (entry.matches(cur, prev, curFut, prevFut)) {
                    holding = true;
                    entryIndex = i;
                    entryPrice = cur.close();
                    entryTs = cur.ts();
                }
                continue;
            }

            boolean lastOverall = (i == n - 1);
            boolean lastBarOfDay = lastOverall || !cur.date().equals(bars.get(i + 1).date());
            ExitDecision d = decideExit(cur, prev, lastOverall, lastBarOfDay, i - entryIndex,
                    exitSpec, exitGroup, entryPrice, curFut, prevFut);
            if (d != null) {
                trades.add(tradeOf(entryTs, cur.ts(), entryPrice, d, null));
                holding = false;
            }
        }

        return aggregate(trades, spec, fees);
    }

    // --------------------------------------------------------------- ETF pre-market mode

    /** Commission-free convenience overload. */
    public BacktestResult runEtfPremarket(StrategySpec spec, BarSeries leverage, BarSeries inverse, BarSeries futures) {
        return runEtfPremarket(spec, leverage, inverse, futures, FeeSchedule.free());
    }

    public BacktestResult runEtfPremarket(StrategySpec spec, BarSeries leverage, BarSeries inverse,
                                          BarSeries futures, FeeSchedule fees) {
        PremarketSpec pm = spec.getPremarket();
        LocalTime start = LocalTime.parse(pm.getStartTime());
        LocalTime end = LocalTime.parse(pm.getEndTime());
        double threshold = Math.abs(pm.getThresholdPct());

        ExitSpec exitSpec = spec.getExit();
        ConditionGroup exitGroup = exitSpec.asGroup();

        Map<LocalDate, List<Bar>> futByDay = groupByDate(futures.bars());
        Map<LocalDate, List<Bar>> levByDay = groupByDate(leverage.bars());
        Map<LocalDate, List<Bar>> invByDay = groupByDate(inverse.bars());

        List<TradeRecord> trades = new ArrayList<>();

        for (Map.Entry<LocalDate, List<Bar>> e : futByDay.entrySet()) {
            Double trend = premarketTrend(e.getValue(), start, end);
            if (trend == null || Math.abs(trend) < threshold) {
                continue; // no clear pre-market trend → skip the day
            }
            boolean up = trend > 0;
            String instrument = up ? "LEVERAGE" : "INVERSE";
            List<Bar> dayBars = (up ? levByDay : invByDay).get(e.getKey());
            if (dayBars == null || dayBars.isEmpty()) {
                continue;
            }
            TradeRecord tr = simulateDay(dayBars, end, exitSpec, exitGroup, instrument);
            if (tr != null) {
                trades.add(tr);
            }
        }

        return aggregate(trades, spec, fees);
    }

    /** Buy the chosen ETF at the first bar at/after the pre-market window ends; exit intraday. */
    private TradeRecord simulateDay(List<Bar> dayBars, LocalTime openAtOrAfter,
                                    ExitSpec exitSpec, ConditionGroup exitGroup, String instrument) {
        int n = dayBars.size();
        int entryIndex = -1;
        for (int i = 0; i < n; i++) {
            if (!dayBars.get(i).ts().toLocalTime().isBefore(openAtOrAfter)) {
                entryIndex = i;
                break;
            }
        }
        if (entryIndex < 0 || entryIndex >= n - 1) {
            return null; // no bar to enter on, or no bar left to exit
        }
        Bar entryBar = dayBars.get(entryIndex);
        double entryPrice = entryBar.close();

        for (int j = entryIndex + 1; j < n; j++) {
            Bar cur = dayBars.get(j);
            Bar prev = dayBars.get(j - 1);
            boolean lastBarOfDay = (j == n - 1);
            // Pre-market mode always closes same day: force day-end at the last bar.
            ExitDecision d = decideExit(cur, prev, false, lastBarOfDay, j - entryIndex,
                    exitSpec, exitGroup, entryPrice, null, null);
            if (d == null && lastBarOfDay) {
                d = new ExitDecision(cur.close(), ExitReason.DAY_END);
            }
            if (d != null) {
                return tradeOf(entryBar.ts(), cur.ts(), entryPrice, d, instrument);
            }
        }
        return null;
    }

    /** Futures close change (%) over [start, end); null if the window has too few bars. */
    private static Double premarketTrend(List<Bar> dayBars, LocalTime start, LocalTime end) {
        Bar first = null;
        Bar last = null;
        for (Bar b : dayBars) {
            LocalTime t = b.ts().toLocalTime();
            if (!t.isBefore(start) && t.isBefore(end)) {
                if (first == null) {
                    first = b;
                }
                last = b;
            }
        }
        if (first == null || last == null || first == last || first.close() == 0.0) {
            return null;
        }
        return (last.close() - first.close()) / first.close() * 100.0;
    }

    // ------------------------------------------------------------------ shared helpers

    /**
     * Stop price for a position entered at {@code entryPrice}, as of wall-clock {@code at} — the
     * time of the bar being checked, so a time band that starts mid-trade moves the level.
     */
    private static double stopPriceOf(ExitSpec exit, double entryPrice, LocalTime at) {
        Double pct = exit.stopLossPctAt(at);
        return pct != null ? entryPrice * (1.0 - pct / 100.0) : Double.NaN;
    }

    private static double tpPriceOf(ExitSpec exit, double entryPrice, LocalTime at) {
        Double pct = exit.takeProfitPctAt(at);
        return pct != null ? entryPrice * (1.0 + pct / 100.0) : Double.NaN;
    }

    private static TradeRecord tradeOf(LocalDateTime entryTs, LocalDateTime exitTs, double entryPrice,
                                       ExitDecision d, String instrument) {
        double ret = (d.price() - entryPrice) / entryPrice * 100.0;
        return new TradeRecord(entryTs, exitTs, entryPrice, d.price(), ret, d.reason(), ret > 0, instrument);
    }

    private static ExitDecision decideExit(Bar cur, Bar prev, boolean lastOverall, boolean lastBarOfDay,
                                           int barsHeld, ExitSpec exit, ConditionGroup exitGroup,
                                           double entryPrice, Bar curFut, Bar prevFut) {
        // Resolved per bar, not per trade: an ExitSpec time band can move these mid-position.
        LocalTime at = cur.ts().toLocalTime();
        double stopPrice = stopPriceOf(exit, entryPrice, at);
        double tpPrice = tpPriceOf(exit, entryPrice, at);
        if (!Double.isNaN(stopPrice) && cur.low() <= stopPrice) {
            return new ExitDecision(stopPrice, ExitReason.STOP_LOSS);
        }
        if (!Double.isNaN(tpPrice) && cur.high() >= tpPrice) {
            return new ExitDecision(tpPrice, ExitReason.TAKE_PROFIT);
        }
        if (!exitGroup.isEmpty() && exitGroup.matches(cur, prev, curFut, prevFut)) {
            return new ExitDecision(cur.close(), ExitReason.SIGNAL);
        }
        if (exit.getMaxHoldBars() != null && barsHeld >= exit.getMaxHoldBars()) {
            return new ExitDecision(cur.close(), ExitReason.TIME);
        }
        if (exit.isCloseAtDayEnd() && lastBarOfDay && !lastOverall) {
            return new ExitDecision(cur.close(), ExitReason.DAY_END);
        }
        if (lastOverall) {
            return new ExitDecision(cur.close(), ExitReason.END_OF_DATA);
        }
        return null;
    }

    private static Map<LocalDate, List<Bar>> groupByDate(List<Bar> bars) {
        Map<LocalDate, List<Bar>> byDay = new LinkedHashMap<>();
        for (Bar b : bars) {
            byDay.computeIfAbsent(b.date(), k -> new ArrayList<>()).add(b);
        }
        return byDay;
    }

    // ------------------------------------------------------------------ aggregation

    private BacktestResult aggregate(List<TradeRecord> priceTrades, StrategySpec spec, FeeSchedule fees) {
        CapitalSpec cap = spec.getCapital() == null ? new CapitalSpec() : spec.getCapital();
        List<TradeRecord> trades = applyCapital(priceTrades, cap, fees);
        int total = trades.size();
        int wins = 0;
        double sumRet = 0.0;
        double sumWin = 0.0;
        double sumLoss = 0.0;
        int lossCount = 0;
        double equity = 1.0;
        double peak = 1.0;
        double maxDrawdown = 0.0;
        int consecLosses = 0;
        int maxConsecLosses = 0;

        List<BacktestResult.EquityPoint> curve = new ArrayList<>();
        for (TradeRecord t : trades) {
            sumRet += t.returnPct();
            if (t.success()) {
                wins++;
                sumWin += t.returnPct();
                consecLosses = 0;
            } else {
                lossCount++;
                sumLoss += t.returnPct();
                consecLosses++;
                maxConsecLosses = Math.max(maxConsecLosses, consecLosses);
            }
            equity *= (1.0 + t.returnPct() / 100.0);
            peak = Math.max(peak, equity);
            double drawdown = (peak - equity) / peak * 100.0;
            maxDrawdown = Math.max(maxDrawdown, drawdown);
            curve.add(new BacktestResult.EquityPoint(t.exitTs(), equity));
        }

        int losses = total - wins;
        double winRate = total == 0 ? 0.0 : (double) wins / total * 100.0;
        double lossRate = total == 0 ? 0.0 : (double) losses / total * 100.0;
        double avgWin = wins == 0 ? 0.0 : sumWin / wins;
        double avgLoss = lossCount == 0 ? 0.0 : sumLoss / lossCount;
        double compounded = (equity - 1.0) * 100.0;

        BacktestResult.Summary summary = new BacktestResult.Summary(
                total, wins, losses, winRate, lossRate, sumRet, compounded,
                avgWin, avgLoss, maxDrawdown, maxConsecLosses);

        BacktestResult.MoneySummary money = moneyOf(trades, cap, fees);
        BacktestResult.FailureAnalysis failure = buildFailureAnalysis(trades);
        BacktestResult.Diagnosis diagnosis =
                FailureDiagnostician.diagnose(summary, money, failure, trades, spec.getExit(), cap,
                        spec.usesPremarket());

        return new BacktestResult(summary, money, diagnosis, trades, curve, failure);
    }

    /**
     * Sizes every trade in won. Shares are whole (Korean equities/ETFs don't trade fractions), so
     * the budget is floored to the nearest share; the leftover cash simply stays uninvested. In
     * compound mode the running balance carries each trade's net result into the next entry.
     */
    private static List<TradeRecord> applyCapital(List<TradeRecord> trades, CapitalSpec cap, FeeSchedule fees) {
        double balance = cap.getAmount();
        List<TradeRecord> out = new ArrayList<>(trades.size());
        for (TradeRecord t : trades) {
            double feeRate = Math.max(0.0, fees.rateFor(t.instrument())) / 100.0;
            double budget = cap.getMode() == CapitalMode.COMPOUND ? balance : cap.getAmount();
            long qty = (t.entryPrice() > 0 && budget > 0) ? (long) Math.floor(budget / t.entryPrice()) : 0L;
            double buyValue = qty * t.entryPrice();
            double sellValue = qty * t.exitPrice();
            double feeCost = (buyValue + sellValue) * feeRate;
            double profit = sellValue - buyValue - feeCost;
            if (cap.getMode() == CapitalMode.COMPOUND) {
                balance += profit;
            }
            out.add(t.withMoney(qty, profit, feeCost));
        }
        return out;
    }

    private static BacktestResult.MoneySummary moneyOf(List<TradeRecord> trades, CapitalSpec cap, FeeSchedule fees) {
        double totalProfit = 0.0;
        double totalFees = 0.0;
        double best = 0.0;
        double worst = 0.0;
        int unaffordable = 0;
        for (TradeRecord t : trades) {
            totalProfit += t.profitAmount();
            totalFees += t.feeAmount();
            best = Math.max(best, t.profitAmount());
            worst = Math.min(worst, t.profitAmount());
            if (t.quantity() == 0) {
                unaffordable++;
            }
        }
        double invest = cap.getAmount();
        double avg = trades.isEmpty() ? 0.0 : totalProfit / trades.size();
        double roc = invest == 0.0 ? 0.0 : totalProfit / invest * 100.0;
        return new BacktestResult.MoneySummary(
                cap.getMode().name(), invest, fees.ratesFor(trades), totalProfit, totalFees,
                avg, invest + totalProfit, roc, best, worst, unaffordable);
    }

    private BacktestResult.FailureAnalysis buildFailureAnalysis(List<TradeRecord> trades) {
        Map<ExitReason, int[]> reasonCount = new EnumMap<>(ExitReason.class);
        Map<ExitReason, double[]> reasonSum = new EnumMap<>(ExitReason.class);
        Map<Integer, int[]> hourLosses = new TreeMap<>();
        Map<Integer, double[]> hourLossSum = new TreeMap<>();

        List<TradeRecord> losing = new ArrayList<>();
        for (TradeRecord t : trades) {
            if (t.success()) {
                continue;
            }
            losing.add(t);
            reasonCount.computeIfAbsent(t.exitReason(), k -> new int[1])[0]++;
            reasonSum.computeIfAbsent(t.exitReason(), k -> new double[1])[0] += t.returnPct();
            hourLosses.computeIfAbsent(t.entryHour(), k -> new int[1])[0]++;
            hourLossSum.computeIfAbsent(t.entryHour(), k -> new double[1])[0] += t.returnPct();
        }

        Map<ExitReason, BacktestResult.ReasonStat> byReason = new EnumMap<>(ExitReason.class);
        for (Map.Entry<ExitReason, int[]> e : reasonCount.entrySet()) {
            int c = e.getValue()[0];
            double sum = reasonSum.get(e.getKey())[0];
            byReason.put(e.getKey(), new BacktestResult.ReasonStat(c, sum / c, sum));
        }

        Map<Integer, BacktestResult.HourStat> byHour = new TreeMap<>();
        for (Map.Entry<Integer, int[]> e : hourLosses.entrySet()) {
            int c = e.getValue()[0];
            double sum = hourLossSum.get(e.getKey())[0];
            byHour.put(e.getKey(), new BacktestResult.HourStat(c, sum / c));
        }

        List<TradeRecord> worst = new ArrayList<>(losing);
        worst.sort((a, b) -> Double.compare(a.returnPct(), b.returnPct()));
        List<TradeRecord> worstTop = worst.subList(0, Math.min(10, worst.size()));

        return new BacktestResult.FailureAnalysis(byReason, byHour, List.copyOf(worstTop));
    }
}

File: backend/src/main/kotlin/com/stockanalysis/backtest/BacktestEngine.kt
package com.stockanalysis.backtest

import com.stockanalysis.backtest.spec.CapitalMode
import com.stockanalysis.backtest.spec.CapitalSpec
import com.stockanalysis.backtest.spec.StrategySpec
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/**
 * 프레임워크에 의존하지 않는 단타 백테스트 엔진.
 *
 * **단일 모드**: 롱 포지션 하나, 시그널이 뜬 봉의 **종가**에 진입합니다(미래를 보지 않습니다).
 * 청산 판정은 [ExitEvaluator]에 맡깁니다.
 *
 * **ETF 장전 모드**([runEtfPremarket]): 매일 장전 선물 변화율이 레버리지/인버스를 고르고
 * ([PremarketDecider]), 고른 ETF를 장 시작에 사서 같은 청산 규칙으로 정리합니다.
 *
 * 진입·청산은 **가격만으로** 정한 뒤, 금액은 [applyCapital]이 별도 패스에서 입힙니다. 그래서
 * 포지션 사이징이나 수수료가 *어떤 거래가 일어나는지*를 바꾸지 않습니다.
 *
 * 판단 두 가지([ExitEvaluator], [PremarketDecider])를 이 클래스가 직접 들고 있지 않은 이유는
 * 실투자가 같은 `StrategySpec`으로 같은 답을 내야 하기 때문입니다.
 */
class BacktestEngine {

    // ------------------------------------------------------------------ 단일 모드

    /** 수수료 없는 편의 오버로드 — 가격만 보는 기준선으로 도구와 테스트가 씁니다. */
    fun run(spec: StrategySpec, series: BarSeries): BacktestResult =
        run(spec, series, FeeSchedule.free())

    fun run(spec: StrategySpec, series: BarSeries, fees: FeeSchedule): BacktestResult {
        val bars = series.bars()
        val n = bars.size
        val trades = mutableListOf<TradeRecord>()

        val entry = spec.entry
        val exitSpec = spec.exit
        val exitGroup = exitSpec.asGroup()

        var holding = false
        var entryIndex = -1
        var entryPrice = 0.0
        var entryTs: LocalDateTime? = null

        for (i in 0 until n) {
            val cur = bars[i]
            val prev = if (i > 0) bars[i - 1] else null
            val curFut = if (series.hasFutures()) series.futuresAt(cur.ts) else null
            val prevFut = if (prev != null && series.hasFutures()) series.futuresAt(prev.ts) else null

            if (!holding) {
                if (entry.matches(cur, prev, curFut, prevFut)) {
                    holding = true
                    entryIndex = i
                    entryPrice = cur.close
                    entryTs = cur.ts
                }
                continue
            }

            val lastOverall = i == n - 1
            val lastBarOfDay = lastOverall || cur.date() != bars[i + 1].date()
            val d = ExitEvaluator.decide(
                cur, prev, lastOverall, lastBarOfDay, i - entryIndex,
                exitSpec, exitGroup, entryPrice, curFut, prevFut
            )
            if (d != null) {
                trades.add(tradeOf(entryTs!!, cur.ts, entryPrice, d, null))
                holding = false
            }
        }

        return aggregate(trades, spec, fees)
    }

    // --------------------------------------------------------------- ETF 장전 모드

    /** 수수료 없는 편의 오버로드. */
    fun runEtfPremarket(
        spec: StrategySpec,
        leverage: BarSeries,
        inverse: BarSeries,
        futures: BarSeries
    ): BacktestResult = runEtfPremarket(spec, leverage, inverse, futures, FeeSchedule.free())

    fun runEtfPremarket(
        spec: StrategySpec,
        leverage: BarSeries,
        inverse: BarSeries,
        futures: BarSeries,
        fees: FeeSchedule
    ): BacktestResult {
        val pm = spec.premarket
        val end = LocalTime.parse(pm.endTime)

        val exitSpec = spec.exit
        val exitGroup = exitSpec.asGroup()

        val futByDay = groupByDate(futures.bars())
        val levByDay = groupByDate(leverage.bars())
        val invByDay = groupByDate(inverse.bars())

        val trades = mutableListOf<TradeRecord>()

        for ((day, dayFutures) in futByDay) {
            val pick = PremarketDecider.decide(PremarketDecider.fromBars(dayFutures), pm)
            if (!pick.shouldTrade()) {
                continue // 장전 추세가 뚜렷하지 않은 날은 건너뜁니다
            }
            val up = pick.side == PremarketDecider.Side.LEVERAGE
            val instrument = pick.instrument()
            val dayBars = (if (up) levByDay else invByDay)[day]
            if (dayBars.isNullOrEmpty()) {
                continue
            }
            simulateDay(dayBars, end, exitSpec, exitGroup, instrument)?.let { trades.add(it) }
        }

        return aggregate(trades, spec, fees)
    }

    /** 장전 구간이 끝난 뒤 첫 봉에 고른 ETF를 사고, 그날 안에 정리합니다. */
    private fun simulateDay(
        dayBars: List<Bar>,
        openAtOrAfter: LocalTime,
        exitSpec: com.stockanalysis.backtest.spec.ExitSpec,
        exitGroup: com.stockanalysis.backtest.spec.ConditionGroup,
        instrument: String?
    ): TradeRecord? {
        val n = dayBars.size
        val entryIndex = dayBars.indexOfFirst { !it.ts.toLocalTime().isBefore(openAtOrAfter) }
        if (entryIndex < 0 || entryIndex >= n - 1) {
            return null // 진입할 봉이 없거나, 청산할 봉이 남지 않았습니다
        }
        val entryBar = dayBars[entryIndex]
        val entryPrice = entryBar.close

        for (j in entryIndex + 1 until n) {
            val cur = dayBars[j]
            val prev = dayBars[j - 1]
            val lastBarOfDay = j == n - 1
            // 장전 모드는 항상 당일 청산이므로 마지막 봉에서 장마감을 강제합니다.
            var d = ExitEvaluator.decide(
                cur, prev, false, lastBarOfDay, j - entryIndex,
                exitSpec, exitGroup, entryPrice, null, null
            )
            if (d == null && lastBarOfDay) {
                d = ExitEvaluator.Decision(cur.close, ExitReason.DAY_END)
            }
            if (d != null) {
                return tradeOf(entryBar.ts, cur.ts, entryPrice, d, instrument)
            }
        }
        return null
    }

    // ------------------------------------------------------------------ 집계

    private fun aggregate(
        priceTrades: List<TradeRecord>,
        spec: StrategySpec,
        fees: FeeSchedule
    ): BacktestResult {
        val cap = spec.capital
        val trades = applyCapital(priceTrades, cap, fees)

        var wins = 0
        var sumRet = 0.0
        var sumWin = 0.0
        var sumLoss = 0.0
        var lossCount = 0
        var equity = 1.0
        var peak = 1.0
        var maxDrawdown = 0.0
        var consecLosses = 0
        var maxConsecLosses = 0

        val curve = mutableListOf<BacktestResult.EquityPoint>()
        for (t in trades) {
            sumRet += t.returnPct
            if (t.success) {
                wins++
                sumWin += t.returnPct
                consecLosses = 0
            } else {
                lossCount++
                sumLoss += t.returnPct
                consecLosses++
                maxConsecLosses = max(maxConsecLosses, consecLosses)
            }
            equity *= (1.0 + t.returnPct / 100.0)
            peak = max(peak, equity)
            maxDrawdown = max(maxDrawdown, (peak - equity) / peak * 100.0)
            curve.add(BacktestResult.EquityPoint(t.exitTs, equity))
        }

        val total = trades.size
        val losses = total - wins
        val summary = BacktestResult.Summary(
            totalTrades = total,
            wins = wins,
            losses = losses,
            winRate = if (total == 0) 0.0 else wins.toDouble() / total * 100.0,
            lossRate = if (total == 0) 0.0 else losses.toDouble() / total * 100.0,
            totalReturnPct = sumRet,
            compoundedReturnPct = (equity - 1.0) * 100.0,
            avgWinPct = if (wins == 0) 0.0 else sumWin / wins,
            avgLossPct = if (lossCount == 0) 0.0 else sumLoss / lossCount,
            maxDrawdownPct = maxDrawdown,
            maxConsecutiveLosses = maxConsecLosses
        )

        val money = moneyOf(trades, cap, fees)
        val failure = buildFailureAnalysis(trades)
        val diagnosis = FailureDiagnostician.diagnose(
            summary, money, failure, trades, spec.exit, cap, spec.usesPremarket()
        )

        return BacktestResult(summary, money, diagnosis, trades, curve, failure)
    }

    /**
     * 모든 거래를 원화로 환산합니다.
     *
     * 한국 주식·ETF는 소수점 주식이 없으므로 예산을 **1주 단위로 내림**하고, 남는 현금은
     * 투자되지 않습니다. 복리 모드에서는 각 거래의 순손익이 다음 진입 예산으로 굴러갑니다.
     */
    private fun applyCapital(
        trades: List<TradeRecord>,
        cap: CapitalSpec,
        fees: FeeSchedule
    ): List<TradeRecord> {
        var balance = cap.amount
        val out = ArrayList<TradeRecord>(trades.size)
        for (t in trades) {
            val feeRate = max(0.0, fees.rateFor(t.instrument)) / 100.0
            val budget = if (cap.mode == CapitalMode.COMPOUND) balance else cap.amount
            val qty = if (t.entryPrice > 0 && budget > 0) floor(budget / t.entryPrice).toLong() else 0L
            val buyValue = qty * t.entryPrice
            val sellValue = qty * t.exitPrice
            val feeCost = (buyValue + sellValue) * feeRate
            val profit = sellValue - buyValue - feeCost
            if (cap.mode == CapitalMode.COMPOUND) {
                balance += profit
            }
            out.add(t.withMoney(qty, profit, feeCost))
        }
        return out
    }

    private fun moneyOf(
        trades: List<TradeRecord>,
        cap: CapitalSpec,
        fees: FeeSchedule
    ): BacktestResult.MoneySummary {
        var totalProfit = 0.0
        var totalFees = 0.0
        var best = 0.0
        var worst = 0.0
        var unaffordable = 0
        for (t in trades) {
            totalProfit += t.profitAmount
            totalFees += t.feeAmount
            best = max(best, t.profitAmount)
            worst = min(worst, t.profitAmount)
            if (t.quantity == 0L) {
                unaffordable++
            }
        }
        val invest = cap.amount
        return BacktestResult.MoneySummary(
            mode = cap.mode.name,
            investAmount = invest,
            feeRatesPct = fees.ratesFor(trades),
            totalProfitAmount = totalProfit,
            totalFeeAmount = totalFees,
            avgProfitPerTrade = if (trades.isEmpty()) 0.0 else totalProfit / trades.size,
            finalBalance = invest + totalProfit,
            returnOnCapitalPct = if (invest == 0.0) 0.0 else totalProfit / invest * 100.0,
            bestTradeAmount = best,
            worstTradeAmount = worst,
            unaffordableTrades = unaffordable
        )
    }

    /** 손실 거래를 청산 사유별·진입 시각별로 묶고, 최악의 거래 10건을 뽑습니다. */
    private fun buildFailureAnalysis(trades: List<TradeRecord>): BacktestResult.FailureAnalysis {
        val losing = trades.filter { !it.success }

        val byReason = losing.groupBy { it.exitReason }
            .mapValues { (_, group) ->
                val sum = group.sumOf { it.returnPct }
                BacktestResult.ReasonStat(group.size, sum / group.size, sum)
            }
            .toSortedMap(compareBy { it.ordinal })

        val byHour = losing.groupBy { it.entryHour() }
            .mapValues { (_, group) ->
                BacktestResult.HourStat(group.size, group.sumOf { it.returnPct } / group.size)
            }
            .toSortedMap()

        val worst = losing.sortedBy { it.returnPct }.take(10)
        return BacktestResult.FailureAnalysis(byReason, byHour, worst)
    }

    private companion object {

        fun tradeOf(
            entryTs: LocalDateTime,
            exitTs: LocalDateTime,
            entryPrice: Double,
            d: ExitEvaluator.Decision,
            instrument: String?
        ): TradeRecord {
            val ret = (d.price - entryPrice) / entryPrice * 100.0
            return TradeRecord(entryTs, exitTs, entryPrice, d.price, ret, d.reason, ret > 0, instrument)
        }

        /** 삽입 순서(=시간순)를 유지해야 하므로 LinkedHashMap 기반으로 묶습니다. */
        fun groupByDate(bars: List<Bar>): Map<LocalDate, List<Bar>> =
            bars.groupByTo(LinkedHashMap()) { it.date() }
    }
}
